packages/react-native-renderer/src/ReactNativeGetListeners.js

```diff
@@ -0,0 +1,168 @@
+
+import {getFiberCurrentPropsFromNode} from './legacy-events/EventPluginUtils';
+import {CustomEvent} from 'react-native/Libraries/ReactPrivate/ReactNativePrivateInterface';
+
+/**
+ * Get a list of listeners for a specific event, in-order.
+ * For React Native we treat the props-based function handlers
+ * as the first-class citizens, and they are always executed first
+ * for both capture and bubbling phase.
+ *
+ * We need "phase" propagated to this point to support the HostComponent
+ * EventEmitter API, which does not mutate the name of the handler based
+ * on phase (whereas prop handlers are registered as `onMyEvent` and `onMyEvent_Capture`).
+ *
+ * Native system events emitted into React Native
+ * will be emitted both to the prop handler function and to imperative event
+ * listeners.
+ *
+ * This will either return null, a single Function without an array, or
+ * an array of 2+ items.
+ */
+export default function getListeners(
+  inst: Fiber,
+  registrationName: string,
+  phase: PropagationPhases,
+  dispatchToImperativeListeners: boolean,
+): null | Function | Array<Function> {
+  // Previously, there was only one possible listener for an event:
+  // the onEventName property in props.
+  // Now, it is also possible to have N listeners
+  // for a specific event on a node. Thus, we accumulate all of the listeners,
+  // including the props listener, and return a function that calls them all in
+  // order, starting with the handler prop and then the listeners in order.
+  // We return either a non-empty array or null.
+  let listeners = null;
+
+  const stateNode = inst.stateNode;
+
+  if (stateNode === null) {
+    return null;
+  }
+
+  // If null: Work in progress (ex: onload events in incremental mode).
+  const props = getFiberCurrentPropsFromNode(stateNode);
+  if (props === null) {
+    // Work in progress.
+    return null;
+  }
+
+  const listener = props[registrationName];
+
+  if (listener && typeof listener !== 'function') {
+    throw new Error(
+      `Expected \`${registrationName}\` listener to be a function, instead got a value of \`${typeof listener}\` type.`,
+    );
+  }
+
+  // If there are no imperative listeners, early exit.
+  if (
+    !(
+      dispatchToImperativeListeners &&
+      stateNode.canonical &&
+      stateNode.canonical._eventListeners
+    )
+  ) {
+    return listener;
+  }
+
+  // TODO: for now, all of these events get an `rn:` prefix to enforce
+  // that the user knows they're only getting non-W3C-compliant events
+  // through this imperative event API.
+  // Events might not necessarily be noncompliant, but we currently have
+  // no verification that /any/ events are compliant.
+  // Thus, we prefix to ensure no collision with W3C event names.
+  const requestedPhaseIsCapture = phase === 'captured';
+  const mangledImperativeRegistrationName = requestedPhaseIsCapture
+    ? 'rn:' + registrationName.replace(/Capture$/, '')
```

Comment:
Alternatively, this could be `invariant(registrationName.endsWith('Capture'), …)`, right?

Reply:
I don't think so (as discussed over VC) - do you agree?

Reply:
The invariant would succeed probably 99% of the time though